feat: add detailed migration guide for Storage (#6750)
aws-amplify/docs: src/pages/[platform]/build-a-backend/storage/storage-v5-to-v6-migration-guide/index.mdx

The input options below have been removed in v6 from Storage API's:

- `provider`: _Custom providers are no longer supported in v6._
- `customPrefix`: _You can [add a prefix resolver](/[platform]/build-a-backend/storage/configure-access/#customize-object-key-path) to your Storage Config to customize the key prefix per accessLevel. See the [Using a Custom Prefix](/[platform]/build-a-backend/storage/storage-v5-to-v6-migration-guide/#using-a-custom-prefix) header for migration details._
- `track`: _Tracking Storage events in Hub is not yet supported in v6._
- `ServerSideEncryption`, `SSECustomerAlgorithm`, `SSECustomerKey`, `SSECustomerKeyMD5`, and `SSEKMSKeyId`: _These options have been deprecated due to low usage and added complexity: Server-Side Encryption has been automatically applied using default keys by S3 [since Jan 2023](https://aws.amazon.com/blogs/aws/amazon-s3-encrypts-new-objects-by-default/)._
- `cacheControl`: _Response payload is always set to 'no-store' in v6._

> - `$metadata`: _Information like `statusCode` and `retryCount` will be provided in v6 via debug messages using [the Console Logger](/[platform]/build-a-backend/utilities/console-logger/)._

Reviewer: Is it already available? If not maybe we can create a public GH feature request and link it here. But it's OK if this doc will be launched same time as logger change.

The `progressCallback` option on upload and download operations has been replaced by `onProgress` in v6. Note that the parameter property names have also changed.

## Storage.put

> The `Storage.put` API has moved to the `uploadData` API in v6. There are a few changes in the experience from v5 to v6 in addition to those mentioned above:

Reviewer 2: ```suggestion
The `Storage.put` API has renamed to `uploadData` API in v6. There are a few changes in the experience from v5 to v6 in addition to those mentioned above:
```
just a suggestion, here and elsewhere if applicable.

Author: Great suggestion, reads much better.

- We have remapped all capitalized input and output keys to camel case.

> - The `cacheControl` option has been removed and the response payload is always set to `no-store` in v6.

Reviewer: This was changed in later revision([link](https://quip-amazon.com/acftAfZE4PGI/Amplify-JS-v6-Storage-API-Rev3#temp:C:LZE3569bab6cb144f50b886f2a65)). 

```suggestion
- The `cacheControl` option has been removed and now defers to the HTTP caching behavior configured by the runtime.
```

> - We removed the `acl` option: the S3 Bucket ACL provides extra access control options that Amplify does not currently support.

Reviewer: Nit: 

```suggestion
- We removed the `acl` option: the S3 Bucket ACL provides extra access control options that Amplify does not currently support. ACL is no longer recommended accroding to [S3 security best practices](https://docs.aws.amazon.com/AmazonS3/latest/userguide/security-best-practices.html#security-best-practices-prevent).
```

- The `expires` option has been deprecated.

> - The `resumable` option has been removed as all upload tasks are resumable in v6. You can find the `pause`, `resume`, and `cancel` functions on the `uploadData` output.

Reviewer: Maybe link to the page mentions the this functions?

        ObjectLockRetainUntilDate?: Date;
        ObjectLockLegalHoldStatus?: ObjectLockLegalHoldStatus | string;
        $metadata: {
          httpStatusCode?: number;
          requestId?: string
        }
      }
      ```

    </div>
    <div>
      **V6**

      ```
      // getUrl
      GetUrlOutput {
        url: URL;
        expiresAt: Date;
      }

      // downloadData
      DownloadDataOutput {
        cancel: (message?: string) => void;
        readonly state: TransferTaskState;
        result: Promise<StorageDownloadDataOutput<Item>>;
      }

      StorageDownloadDataOutput<Item> {
        key: string;
        lastModified?: Date;
        size?: number;
        eTag?: string;
        metadata?: Record<string, string>;
        versionId?: string;
        contentType?: string;
        body: {
          blob: () => Promise<Blob>;
          json: () => Promise<any>;
          text: () => Promise<string>;
        };
      }
      ```

    </div>
  </Columns>
</Accordion>

### Get URL Only (Default)

<BlockSwitcher>
  <Block name="V6">
    ```ts
    import { getUrl } from 'aws-amplify/storage';

    const handleGetUrl = async (key: string) => {

>       // Generate a file download url with check if the file exists in the S3 bucket

Reviewer: Nit:

```suggestion
      // Check if the file exists in the S3 bucket and generate a file download url.

        nextToken?: string;
        hasNextToken: boolean;
      };

      S3ProviderListOutputItem {
        key: string;
        eTag: string;
        lastModified: Date;
        size: number;
      }
      ```

    </div>
    <div>
      **V6**

      ```
      // Paginated List
      ListPaginateOutput {
        items: ListOutputItem[];
        nextToken?: string;
      }

      // List All
      ListAllOutput {
        items: ListOutputItem[];
      }

      ListOutputItem {
        key: string;
        lastModified?: Date;
        size?: number;
        eTag?: string;
      }
      ```

    </div>
  </Columns>
</Accordion>

### Paginated List (Default)

<BlockSwitcher>
  <Block name="V6">
    ```ts
    import { list } from 'aws-amplify/storage';

    const handleList = async (
      prefix: string, 
      accessLevel: string,
      id: string
    ) => {
      const result = Storage.list({
        prefix,
        options: {
          // accessLevel and targetIdentityId not required if accessLevel is 'guest'
          accessLevel,
          targetIdentityId: id
        }
      });
      const { items } = result;
      const nextToken = result.nextToken || null;
    }

>     ```

Reviewer: This is not the best code snippet to show the pagination. I'd recommend a code snippet actually make make multiple list API calls with `nextToken`. Similar to [v5 example](https://docs.amplify.aws/javascript/prev/build-a-backend/storage/list/#paginated-file-access)

Author: Good idea, can do.

  </Block>
  <Block name="V5">
    ```ts
    import { Storage } from 'aws-amplify';

>     const handleList = async (

Reviewer: Ditto, I'd suggest use the v5 example in the list section.

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'Migrate from v5 to v6',
  description: 'Learn more about the migration steps to upgrade Storage APIs for Amplify JavaScript v5 to v6',
  platforms: ['angular', 'javascript', 'nextjs', 'react', 'vue', 'react-native']
};

export const getStaticPaths = async () => {
  return getCustomStaticPath(meta.platforms);
};

export function getStaticProps(context) {
  return {
    props: {
      platform: context.params.platform,
      meta
    }
  };
}

> This guide will help you migrate v5 Storage implementations to the new functional API's in v6.

Author: AH good catch! All my life I've been doing it wrong, I guess you only add apostrophes to acronyms if they include periods. TIL!

The input options below have been removed in v6 from Storage API's:

- `provider`: _Custom providers are no longer supported in v6._
- `customPrefix`: _You can [add a prefix resolver](/[platform]/build-a-backend/storage/configure-access/#customize-object-key-path) to your Storage Config to customize the key prefix per accessLevel. See the [Using a Custom Prefix](/[platform]/build-a-backend/storage/storage-v5-to-v6-migration-guide/#using-a-custom-prefix) header for migration details._
- `track`: _Tracking Storage events in Hub is not yet supported in v6._
- `ServerSideEncryption`, `SSECustomerAlgorithm`, `SSECustomerKey`, `SSECustomerKeyMD5`, and `SSEKMSKeyId`: _These options have been deprecated due to low usage and added complexity: Server-Side Encryption has been automatically applied using default keys by S3 [since Jan 2023](https://aws.amazon.com/blogs/aws/amazon-s3-encrypts-new-objects-by-default/)._
- `cacheControl`: _This option has been removed and now defers to the HTTP caching behavior configured by the runtime._

> - `$metadata`: _Information like `httpStatusCode` and `attempts` will be provided in an upcoming release via debug messages using [the Console Logger](/[platform]/build-a-backend/utilities/console-logger/)._

Reviewer 3: I think this might not be tied specifically to the Console Logger but a Logging category/feature

        useAccelerateEndpoint?: boolean;
        resumable?: boolean;
      }
      // resumable config
      config?:  {
        // All of the standard config above plus
        resumable: true; // More specific than resumable above
        progressCallback?: (progress: {
          loaded: number;
          total: number;
        }) => any; // More specific than progressCallback above
        completeCallback?: (event: { key?: string }) => any;
        errorCallback?: (err: any) => any;
      }
      ```

    </div>
    <div>
      **V6**

      ```
      input: {
        key: string;
        data: string | Blob | ArrayBufferView | ArrayBuffer;
        options?: {
          accessLevel?: any;
          useAccelerateEndpoint?: boolean;
          onProgress?: (event: {
            transferredBytes: number;
            totalBytes?: number;
          }) => void;
          contentDisposition?: string;
          contentEncoding?: string;
          contentType?: string;
          metadata?: Record<string, string>;
        };
      }
      ```
    </div>
  </Columns>
</Accordion>

<Accordion title='Output' headingLevel='3'>
  <Columns columns={2}>
    <div>
      **V5**

      ```

>       S3ProviderGetPropertiesOutput {

Reviewer 3: QQ: The output in v5 `put` was a getProperties one? 🤯

Author: Oh I'm wondering if I accidentally edited the wrong one when I was adding the getProperties example? I will take a look, good catch.

<Accordion title='Input' headingLevel='3'>
  <Columns columns={2}>
    <div>
      **V5**

      ```
      // Assumes default provider 'AWSS3': custom provider not supported in v6
      key: string
      options: {
        level?: 'public' | 'protected' | 'private';
        provider?: 'AWSS3';
        customPrefix?: {
          private?: string;
          public?: string;
          protected?: string;
        };
        track?: boolean; // Not supported in v6
        download?: boolean; // Not supported in v6
        expires?: number;
        identityId?: string;
        progressCallback?: (progress: any) => any;
        cacheControl?: string;
        contentDisposition?: string; // Not supported in v6
        contentEncoding?: string; // Not supported in v6
        contentLanguage?: string; // Not supported in v6
        contentType?: string; // Not supported in v6
        SSECustomerAlgorithm?: string;
        SSECustomerKey?: string;
        SSECustomerKeyMD5?: string;
        validateObjectExistence?: boolean;
      };
      ```

    </div>
    <div>
      **V6**

      ```
      // getUrl
      {
        key: string;
        options?: {

>           accessLevel?: 'guest' | 'private' | 'protected';

Reviewer: Super nit: to map with v5 snippet:
```suggestion
          accessLevel?: 'guest' | 'protected' | 'private';
```

    ```ts
    import { Storage } from 'aws-amplify';

    const handleDownload = async (key: string) => {
      const { Body, ETag } = await Storage.get(key, {
        download: true,
        validateObjectExistence: true,
        progressCallback: ({ loaded, total }) => {
          // Progress implementation
        }
      });
    }
    ```

  </Block>
</BlockSwitcher>

### Monitor Download Progress

The `progressCallback` option has been replaced by `onProgress` in v6. The function input property names `loaded` and `total` now map to `transferredBytes` and `totalBytes`.

<BlockSwitcher>
  <Block name="V6">
  ```
  import { downloadData } from 'aws-amplify/storage';

  const handleDownload = async (key: string) => {
    const { body, eTag } = await downloadData({
      key,
      options: {
        accessLevel: 'guest',
        onProgress: ({ transferredBytes, totalBytes }) => {
          // Progress implementation
        }
      }
    }).result;
  }
  ```

  </Block>
  <Block name="V5">
    ```
    import { Storage } from 'aws-amplify';

    const handleDownload = async (key: string) => {
      const { Body, ETag } = await Storage.get(key, {
        download: true,

>         validateObjectExistence: true

Reviewer: Is this option extra?

Author: Yeah, basically unnecessary, wanted to demonstrate other parameters and how they map.

      });
    }
    ```

  </Block>
</BlockSwitcher>

## Storage.copy

In addition to [the changes mentioned above](/[platform]/build-a-backend/storage/storage-v5-to-v6-migration-guide/#deprecated-options), the `copy` API in v6 no longer supports content options (`contentDisposition`, `contentLanguage`, and `contentType`) as they override options selected on the original file `upload` and have caused complexity in implementation.

<Accordion title='Input' headingLevel='3'>
  <Columns columns={2}>
    <div>
      **V5**

      ```
      // Assumes default provider 'AWSS3': custom provider not supported in v6
      src: {
        key: string;
        level?: string;
        identityId?: string;
      }
      dest: {
        level?: string;
        key: string;
      }
      config?: {
        provider?: 'AWSS3';
        customPrefix?: {
          public?: string;
          protected?: string;
          private?: string;
        };

>         track?: boolean;

Reviewer 3: Are all these still supported in v6? I noticed these options had comments in APIs above - should they be here also?

Author: Good catch, forgot to move those comments down and apply across the board.

        requestId?: string;
        extendedRequestId?: string;
        cfId?: string;
        attempts?: number;
        totalRetryDelay?: number;
      };
      ```

    </div>
    <div>
      **V6**

      ```
      RemoveOutput {
        key: string;
      }
      ```

    </div>
  </Columns>
</Accordion>

<BlockSwitcher>
  <Block name="V6">
    ```ts
    import { remove } from 'aws-amplify/storage';

    const handleRemove = async (key: string, accessLevel: string) => {
      await remove({ 
        key,
        options: {
          // accessLevel is required if other than 'guest'
          accessLevel
        }
      });
    }
    ```

  </Block>
  <Block name="V5">
    ```ts
    import { Storage } from 'aws-amplify';

    const handleRemove = async (key: string, accessLevel: string) => {
      await Storage.remove(key, {
        // accessLevel is required if other than 'public'
        level: accessLevel
      });
    }
    ```

  </Block>
</BlockSwitcher>

## Storage.list

There are a few changes in the `list` API from v5 to v6 in addition to [those mentioned above](/[platform]/build-a-backend/storage/storage-v5-to-v6-migration-guide/#deprecated-options):

> - The `path` parameter is now optional. If you want to list all files in the public folder, you do not need to provide an empty string as the `path` argument.

Reviewer: `path` in v5 is replaced by `prefix` now, instead of made optional.

        lastModified: Date;
        metadata: Record<string, string>;
      };
      ```

    </div>
    <div>
      **V6**

      ```
      {
        key: string;
        lastModified?: Date;
        size?: number';
        eTag?: string';
        metadata?: Record<string, string>';
        versionId?: string;
        contentType?: string;
      }
      ```

    </div>
  </Columns>
</Accordion>

<BlockSwitcher>
  <Block name="V6">
    ```ts
    import { getProperties } from 'aws-amplify/storage';

    const handleGetProperties = async (
      key: string,
      accessLevel: string
    ) => {
      const result = await getProperties({
        key: 'key',
        options: {
          // accessLevel not required if accessLevel is 'guest'
          accessLevel,
        }
      });
    }
    ```

  </Block>
  <Block name="V5">
    ```ts
    import { Storage } from 'aws-amplify';

    const handleGetProperties = async (
      key: string,
      accessLevel: string
    ) => {
      const result = await Storage.getProperties(key, {
        // level not required if level is 'public'
        level: accessLevel,
      });
    }
    ```

  </Block>
</BlockSwitcher>

## Storage.cancel

> The process for cancelling a request has changed in v6. In v5 you would send in the promise or `UploadTask` as input to the `Storage.cancel` API. In v6, `cancel` is a function returned with the result of a Storage operation. To cancel an operation, you will call `operation.cancel()`.

Reviewer: ```suggestion
The process for cancelling a request has changed in v6. In v5, you have to maintain the promise reference returned from other storage API calls, and supply to `Storage.cancel` API. In v6, `cancel` is a function returned with the result of a Storage operation. To cancel an operation, you will call `operation.cancel()`.
```

  };
}

This guide will help you migrate v5 Storage implementations to the new functional APIs in v6.

## Deprecated Options

<Callout>Many input and output properties have been deprecated due to low usage or added complexity. If you would like to see these options re-implemented in v6, feel free to open a [feature request](https://github.com/aws-amplify/amplify-js/issues/new?assignees=&labels=pending-triage&projects=&template=2.feature_request.yaml) in our github repository.</Callout>

The input options below have been removed in v6 from Storage APIs:

> - `provider`: _Custom providers are no longer supported in v6. All examples below use the input/output values expected from the Default AWSS3Provider._

Reviewer 3: Nit(?): This seems to be saying that v6 samples are using AWSS3Provider but I think we just refer to it as the S3 Provider in v6

- `customPrefix`: _You can [add a prefix resolver](/[platform]/build-a-backend/storage/configure-access/#customize-object-key-path) to your Storage Config to customize the key prefix per accessLevel. See the [Using a Custom Prefix](/[platform]/build-a-backend/storage/storage-v5-to-v6-migration-guide/#using-a-custom-prefix) header for migration details._

> - `track`: _Auto-tracking Storage events in pinpoint is Deprecated in V6. You can track Storage events [using the Analytics category](/[platform]/build-a-backend/more-features/analytics/record-events/)._

Reviewer 3: Nit: Should we reuse the language "has been removed"? Deprecated usually implies it's still around but no longer supported/recommended

Author: Yes, has been removed overall is better.

- `ServerSideEncryption`, `SSECustomerAlgorithm`, `SSECustomerKey`, `SSECustomerKeyMD5`, and `SSEKMSKeyId`: _These options have been deprecated due to low usage and added complexity: Server-Side Encryption has been automatically applied using default keys by S3 [since Jan 2023](https://aws.amazon.com/blogs/aws/amazon-s3-encrypts-new-objects-by-default/)._
- `cacheControl`: _This option has been removed and now defers to the HTTP caching behavior configured by the runtime._

> - `$metadata`: _Information like `httpStatusCode` and `attempts` will be provided in an upcoming release via category-level debug logs._

Reviewer 3: Still reads a little off to me. Suggestion:
```suggestion
- `$metadata`: _Debugging information like `httpStatusCode` and `attempts` will be reintroduced in an upcoming release focused on changes to logging._
```

  };
}

This guide will help you migrate Amplify JavaScript v5's Storage APIs to the new Amplify JavaScript v6's Storage APIs.

## Deprecated Options

> <Callout>Many input and output properties have been deprecated in favor of alternatives with a better developer experience. If you have identified a use case that you feel isn't covered in this migration guide, feel free to open a [GitHub issue](https://github.com/aws-amplify/amplify-js/issues/new?assignees=&labels=pending-triage&projects=&template=2.feature_request.yaml).</Callout>

Reviewer 3: Small nit: I wonder if we should just say "please open a" vs "feel free to open a". The latter can sometimes come off a little sarcastic depending on interpretation?

Reviewer 4: adddressed.

- `$metadata`: _Information like `httpStatusCode` and `attempts` will be provided in an upcoming release via debug messages using [the Console Logger](/[platform]/build-a-backend/utilities/console-logger/)._
- `bucket`: _The `bucket` is required for `Storage` configuration object as part of `Amplify.configure()` and has therefore been removed from input options._

>Note: We have remapped all capitalized input and output keys to "camelCase" in Amplify JavaScript v6.

## Storage.put

The `Storage.put` API has been renamed to `uploadData` in v6. There are a few changes in the experience from v5 to v6 in addition to [those mentioned above](/[platform]/build-a-backend/storage/storage-v5-to-v6-migration-guide/#deprecated-options):

- We removed the `ACL` option: the S3 Bucket ACL provides extra access control options that Amplify does not currently support. ACL is no longer recommended according to [S3 security best practices](https://docs.aws.amazon.com/AmazonS3/latest/userguide/security-best-practices.html#security-best-practices-prevent).
- The `Expires` option has been deprecated.

> - The `resumable` option has been removed from `config` as all upload tasks are resumable by-default in Amplify JavaScript v6. You can find the `pause`, `resume`, and `cancel` functions on `uploadData`'s return value. See [the v6 Storage documentation](/[platform]/build-a-backend/storage/upload/#pause-and-resume-upload) for details.

Reviewer 3: Small nit: I don't think "by default" needs to be hyphenated here

      // AWSS3 (Default) Provider
      key: string
      object: {

>         ACL?: string; // Deprecated in V6

Reviewer 3: Nit: Kind of feel we should err towards "removed" instead of deprecated for the same reason?
